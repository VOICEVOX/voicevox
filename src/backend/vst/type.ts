import { TrackId } from "@/type/preload";

export type Routing = {
  channelMode: "mono" | "stereo";
  channelIndex: Record<TrackId, number>;
};
