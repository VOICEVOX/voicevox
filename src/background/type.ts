import { z } from "zod";
import { engineIdSchema } from "@/type/preload";

// デフォルトエンジン情報
export const defaultEngineInfosEnvSchema = z
  .object({
    uuid: engineIdSchema,
    host: z.string(),
    name: z.string(),
    executionEnabled: z.boolean(),
    executionFilePath: z.string(),
    executionArgs: z.array(z.string()),
    path: z.string().optional(),
  })
  .array();
