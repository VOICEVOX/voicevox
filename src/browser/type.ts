import { IpcIHData } from "@/type/ipc";

export type WorkerToMainMessage = {
  [K in keyof IpcIHData]: {
    type: K;
    return: IpcIHData[K]["return"];
    eventId: string;
  };
}[keyof IpcIHData];

export type MainToWorkerMessage = {
  [K in keyof IpcIHData]: {
    type: K;
    args: IpcIHData[K]["args"];
    eventId: string;
  };
}[keyof IpcIHData];
