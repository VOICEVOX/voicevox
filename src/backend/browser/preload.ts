import { SandboxKey, Sandbox } from "../type/preload";
import { api } from "./sandbox";

const sandbox: Sandbox = api;
// @ts-expect-error readonlyになっているが、初期化処理はここで行うので問題ない
window[SandboxKey] = sandbox;
