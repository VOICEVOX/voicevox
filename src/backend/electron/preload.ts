import { contextBridge, ipcRenderer } from "electron";
import type { IpcRendererInvoke } from "./ipc";
import {
  ConfigType,
  EngineId,
  Sandbox,
  SandboxKey,
  TextAsset,
} from "@/type/preload";

const ipcRendererInvokeProxy = new Proxy(
  {},
  {
    get:
      (_, channel: string) =>
      (...args: unknown[]) =>
        ipcRenderer.invoke(channel, ...args),
  },
) as IpcRendererInvoke;

const api: Sandbox = {
  getAppInfos: async () => {
    return await ipcRendererInvokeProxy.GET_APP_INFOS();
  },

  getTextAsset: (textType) => {
    return ipcRendererInvokeProxy.GET_TEXT_ASSET(textType) as Promise<
      TextAsset[typeof textType]
    >;
  },

  getAltPortInfos: async () => {
    return await ipcRendererInvokeProxy.GET_ALT_PORT_INFOS();
  },

  showAudioSaveDialog: ({ title, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_AUDIO_SAVE_DIALOG({
      title,
      defaultPath,
    });
  },

  showTextSaveDialog: ({ title, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_TEXT_SAVE_DIALOG({ title, defaultPath });
  },

  showSaveDirectoryDialog: ({ title }) => {
    return ipcRendererInvokeProxy.SHOW_SAVE_DIRECTORY_DIALOG({ title });
  },

  showVvppOpenDialog: ({ title, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_VVPP_OPEN_DIALOG({ title, defaultPath });
  },

  showOpenDirectoryDialog: ({ title }) => {
    return ipcRendererInvokeProxy.SHOW_OPEN_DIRECTORY_DIALOG({ title });
  },

  showProjectSaveDialog: ({ title, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_PROJECT_SAVE_DIALOG({
      title,
      defaultPath,
    });
  },

  showProjectLoadDialog: ({ title }) => {
    return ipcRendererInvokeProxy.SHOW_PROJECT_LOAD_DIALOG({ title });
  },

  showImportFileDialog: ({ title, name, extensions }) => {
    return ipcRendererInvokeProxy.SHOW_IMPORT_FILE_DIALOG({
      title,
      name,
      extensions,
    });
  },

  writeFile: async ({ filePath, buffer }) => {
    return await ipcRendererInvokeProxy.WRITE_FILE({ filePath, buffer });
  },

  readFile: async ({ filePath }) => {
    return await ipcRendererInvokeProxy.READ_FILE({ filePath });
  },

  isAvailableGPUMode: () => {
    return ipcRendererInvokeProxy.IS_AVAILABLE_GPU_MODE();
  },

  isMaximizedWindow: () => {
    return ipcRendererInvokeProxy.IS_MAXIMIZED_WINDOW();
  },

  onReceivedIPCMsg: (listeners) => {
    Object.entries(listeners).forEach(([channel, listener]) => {
      ipcRenderer.on(channel, listener);
    });
  },

  closeWindow: () => {
    void ipcRendererInvokeProxy.CLOSE_WINDOW();
  },

  minimizeWindow: () => {
    void ipcRendererInvokeProxy.MINIMIZE_WINDOW();
  },

  maximizeWindow: () => {
    void ipcRendererInvokeProxy.MAXIMIZE_WINDOW();
  },

  logError: (...params) => {
    console.error(...params);
    // 経緯 https://github.com/VOICEVOX/voicevox/pull/1620#discussion_r1371804569
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "error",
    });
  },

  logWarn: (...params) => {
    console.warn(...params);
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "warn",
    });
  },

  logInfo: (...params) => {
    console.info(...params);
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "info",
    });
  },

  openLogDirectory: () => {
    void ipcRendererInvokeProxy.OPEN_LOG_DIRECTORY();
  },

  engineInfos: () => {
    return ipcRendererInvokeProxy.ENGINE_INFOS();
  },

  restartEngine: (engineId: EngineId) => {
    return ipcRendererInvokeProxy.RESTART_ENGINE({ engineId });
  },

  openEngineDirectory: (engineId: EngineId) => {
    return ipcRendererInvokeProxy.OPEN_ENGINE_DIRECTORY({ engineId });
  },

  checkFileExists: (file) => {
    return ipcRendererInvokeProxy.CHECK_FILE_EXISTS({ file });
  },

  changePinWindow: () => {
    void ipcRendererInvokeProxy.CHANGE_PIN_WINDOW();
  },

  hotkeySettings: (newData) => {
    return ipcRendererInvokeProxy.HOTKEY_SETTINGS({ newData });
  },

  getDefaultHotkeySettings: async () => {
    return await ipcRendererInvokeProxy.GET_DEFAULT_HOTKEY_SETTINGS();
  },

  getDefaultToolbarSetting: async () => {
    return await ipcRendererInvokeProxy.GET_DEFAULT_TOOLBAR_SETTING();
  },

  setNativeTheme: (source) => {
    void ipcRendererInvokeProxy.SET_NATIVE_THEME(source);
  },

  vuexReady: () => {
    void ipcRendererInvokeProxy.ON_VUEX_READY();
  },

  /**
   * 設定情報を取得する
   */
  getSetting: async (key) => {
    return (await ipcRendererInvokeProxy.GET_SETTING(
      key,
    )) as ConfigType[typeof key];
  },

  /**
   * 設定情報を保存する
   */
  setSetting: async (key, newValue) => {
    return (await ipcRendererInvokeProxy.SET_SETTING(
      key,
      newValue,
    )) as typeof newValue;
  },

  setEngineSetting: async (engineId, engineSetting) => {
    await ipcRendererInvokeProxy.SET_ENGINE_SETTING(engineId, engineSetting);
  },

  installVvppEngine: async (filePath) => {
    return await ipcRendererInvokeProxy.INSTALL_VVPP_ENGINE(filePath);
  },

  uninstallVvppEngine: async (engineId) => {
    return await ipcRendererInvokeProxy.UNINSTALL_VVPP_ENGINE(engineId);
  },

  validateEngineDir: async (engineDir) => {
    return await ipcRendererInvokeProxy.VALIDATE_ENGINE_DIR({ engineDir });
  },

  /**
   * アプリを再読み込みする。
   * 画面以外の情報を刷新する。
   */
  reloadApp: async ({ isMultiEngineOffMode }) => {
    await ipcRendererInvokeProxy.RELOAD_APP({ isMultiEngineOffMode });
  },
};

contextBridge.exposeInMainWorld(SandboxKey, api);
