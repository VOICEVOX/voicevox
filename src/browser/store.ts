import { defaultEngine, directoryHandleStoreKey } from "./contract";
import {
  EngineId,
  electronStoreSchema,
  engineSettingSchema,
} from "@/type/preload";

const dbName = "voicevox-web";
// FIXME: DBのschemaを変更したら、dbVersionを上げる
// TODO: 気づけるようにしたい
const dbVersion = 1;
// NOTE: settingを複数持つことはないと仮定して、keyを固定してしまう
export const entryKey = "value";

export const openDB = () =>
  new Promise<IDBDatabase>((resolve, reject) => {
    const request = indexedDB.open(dbName, dbVersion);
    request.onsuccess = () => {
      resolve(request.result);
    };
    request.onerror = () => {
      // TODO: handling
      reject(request.error);
    };
    request.onupgradeneeded = (ev) => {
      if (ev.oldVersion === 0) {
        // Initialize
        const db = request.result;
        const baseSchema = electronStoreSchema.parse({});
        Object.entries(baseSchema).forEach(([key, value]) => {
          const k = key as keyof typeof baseSchema;
          if (k !== "engineSettings") {
            db.createObjectStore(key).add(value, entryKey);
            return;
          }
          // defaultのEngineSettingを追加
          const defaultVoicevoxEngineId = EngineId(defaultEngine.uuid);
          db.createObjectStore(key).add(
            {
              [defaultVoicevoxEngineId]: engineSettingSchema.parse({}),
            },
            entryKey
          );
        });

        // DirectoryHandleも格納する
        db.createObjectStore(directoryHandleStoreKey);
      }
      // TODO: migrate
    };
  });
