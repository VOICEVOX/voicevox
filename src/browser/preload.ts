import { SandboxKey } from "../type/preload";
import { api } from "./sandbox";

// @ts-expect-error readonlyになっているが、初期化処理はここで行うので問題ない
window[SandboxKey] = api;
