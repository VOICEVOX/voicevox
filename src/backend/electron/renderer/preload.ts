import { contextBridge, ipcRenderer, webUtils } from "electron";
import {
  wrapToTransferableResult,
  getOrThrowTransferableResult,
} from "../transferableResultHelper";
import { type IpcRendererInvoke } from "./ipc";
import { BridgeKey, SandboxWithTransferableResult } from "./backendApiLoader";
import { ConfigType, EngineId, Sandbox, TextAsset } from "@/type/preload";

const ipcRendererInvokeProxy = new Proxy(
  {},
  {
    get:
      (_, channel: string) =>
      async (...args: unknown[]) => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        const transferableResult = await ipcRenderer.invoke(channel, ...args);
        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
        return getOrThrowTransferableResult(transferableResult);
      },
  },
) as IpcRendererInvoke;

const api: Sandbox = {
  getTextAsset: (textType) => {
    return ipcRendererInvokeProxy.GET_TEXT_ASSET(textType) as Promise<
      TextAsset[typeof textType]
    >;
  },

  getAltPortInfos: async () => {
    return await ipcRendererInvokeProxy.GET_ALT_PORT_INFOS();
  },

  getInitialProjectFilePath: async () => {
    return await ipcRendererInvokeProxy.GET_INITIAL_PROJECT_FILE_PATH();
  },

  showSaveDirectoryDialog: ({ title }) => {
    return ipcRendererInvokeProxy.SHOW_SAVE_DIRECTORY_DIALOG({ title });
  },

  showOpenDirectoryDialog: ({ title }) => {
    return ipcRendererInvokeProxy.SHOW_OPEN_DIRECTORY_DIALOG({ title });
  },

  showOpenFileDialog: ({ title, name, extensions, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_OPEN_FILE_DIALOG({
      title,
      name,
      extensions,
      defaultPath,
    });
  },

  showSaveFileDialog: ({ title, name, extensions, defaultPath }) => {
    return ipcRendererInvokeProxy.SHOW_SAVE_FILE_DIALOG({
      title,
      name,
      extensions,
      defaultPath,
    });
  },

  writeFile: async ({ filePath, buffer }) => {
    return await ipcRendererInvokeProxy.WRITE_FILE({ filePath, buffer });
  },

  readFile: async ({ filePath }) => {
    return await ipcRendererInvokeProxy.READ_FILE({ filePath });
  },

  isAvailableGPUMode: () => {
    return ipcRendererInvokeProxy.IS_AVAILABLE_GPU_MODE();
  },

  isMaximizedWindow: () => {
    return ipcRendererInvokeProxy.IS_MAXIMIZED_WINDOW();
  },

  onReceivedIPCMsg: (listeners) => {
    Object.entries(listeners).forEach(([channel, listener]) => {
      ipcRenderer.on(channel, listener);
    });
  },

  closeWindow: () => {
    void ipcRendererInvokeProxy.CLOSE_WINDOW();
  },

  minimizeWindow: () => {
    void ipcRendererInvokeProxy.MINIMIZE_WINDOW();
  },

  toggleMaximizeWindow: () => {
    void ipcRendererInvokeProxy.TOGGLE_MAXIMIZE_WINDOW();
  },

  toggleFullScreen: () => {
    void ipcRendererInvokeProxy.TOGGLE_FULLSCREEN();
  },

  zoomIn: () => {
    void ipcRendererInvokeProxy.ZOOM_IN();
  },
  zoomOut: () => {
    void ipcRendererInvokeProxy.ZOOM_OUT();
  },
  zoomReset: () => {
    void ipcRendererInvokeProxy.ZOOM_RESET();
  },

  logError: (...params) => {
    console.error(...params);
    // 経緯 https://github.com/VOICEVOX/voicevox/pull/1620#discussion_r1371804569
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "error",
    });
  },

  logWarn: (...params) => {
    console.warn(...params);
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "warn",
    });
  },

  logInfo: (...params) => {
    console.info(...params);
    ipcRenderer.send("__ELECTRON_LOG__", {
      data: [...params],
      level: "info",
    });
  },

  openLogDirectory: () => {
    void ipcRendererInvokeProxy.OPEN_LOG_DIRECTORY();
  },

  engineInfos: () => {
    return ipcRendererInvokeProxy.ENGINE_INFOS();
  },

  restartEngine: (engineId: EngineId) => {
    return ipcRendererInvokeProxy.RESTART_ENGINE({ engineId });
  },

  openEngineDirectory: (engineId: EngineId) => {
    return ipcRendererInvokeProxy.OPEN_ENGINE_DIRECTORY({ engineId });
  },

  checkFileExists: (file) => {
    return ipcRendererInvokeProxy.CHECK_FILE_EXISTS({ file });
  },

  changePinWindow: () => {
    void ipcRendererInvokeProxy.CHANGE_PIN_WINDOW();
  },

  hotkeySettings: (newData) => {
    return ipcRendererInvokeProxy.HOTKEY_SETTINGS({ newData });
  },

  getDefaultToolbarSetting: async () => {
    return await ipcRendererInvokeProxy.GET_DEFAULT_TOOLBAR_SETTING();
  },

  setNativeTheme: (source) => {
    void ipcRendererInvokeProxy.SET_NATIVE_THEME(source);
  },

  vuexReady: () => {
    void ipcRendererInvokeProxy.ON_VUEX_READY();
  },

  /**
   * 設定情報を取得する
   */
  getSetting: async (key) => {
    return (await ipcRendererInvokeProxy.GET_SETTING(
      key,
    )) as ConfigType[typeof key];
  },

  /**
   * 設定情報を保存する
   */
  setSetting: async (key, newValue) => {
    return (await ipcRendererInvokeProxy.SET_SETTING(
      key,
      newValue,
    )) as typeof newValue;
  },

  setEngineSetting: async (engineId, engineSetting) => {
    await ipcRendererInvokeProxy.SET_ENGINE_SETTING(engineId, engineSetting);
  },

  installVvppEngine: async (filePath) => {
    return await ipcRendererInvokeProxy.INSTALL_VVPP_ENGINE(filePath);
  },

  uninstallVvppEngine: async (engineId) => {
    return await ipcRendererInvokeProxy.UNINSTALL_VVPP_ENGINE(engineId);
  },

  validateEngineDir: async (engineDir) => {
    return await ipcRendererInvokeProxy.VALIDATE_ENGINE_DIR({ engineDir });
  },

  /**
   * アプリを再読み込みする。
   * 画面以外の情報を刷新する。
   */
  reloadApp: async ({ isMultiEngineOffMode }) => {
    await ipcRendererInvokeProxy.RELOAD_APP({ isMultiEngineOffMode });
  },

  /** webUtils.getPathForFileを呼ぶ */
  getPathForFile: (file) => {
    return webUtils.getPathForFile(file);
  },
};

const wrapApi = (baseApi: Sandbox): SandboxWithTransferableResult => {
  const wrappedApi = {} as SandboxWithTransferableResult;
  for (const key in baseApi) {
    const propKey = key as keyof SandboxWithTransferableResult;
    // @ts-expect-error とりあえず動くので無視
    wrappedApi[propKey] = async (...args: unknown[]) => {
      // @ts-expect-error とりあえず動くので無視
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
      return wrapToTransferableResult(() => baseApi[propKey](...args));
    };
  }
  return wrappedApi;
};
contextBridge.exposeInMainWorld(BridgeKey, wrapApi(api));
