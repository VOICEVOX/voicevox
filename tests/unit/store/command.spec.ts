import { store } from "@/store";
import { AudioKey } from "@/type/preload";
import { resetMockMode, uuid4 } from "@/helpers/random";
import { cloneWithUnwrapProxy } from "@/helpers/cloneWithUnwrapProxy";

const initialState = cloneWithUnwrapProxy(store.state);
beforeEach(() => {
  store.replaceState(initialState);

  resetMockMode();
});

test("コマンド実行で履歴が作られる", async () => {
  await store.dispatch("COMMAND_SET_AUDIO_KEYS", {
    audioKeys: [AudioKey(uuid4())],
  });
  const { audioKeys, redoCommands, undoCommands } = store.state;
  expect({ audioKeys, redoCommands, undoCommands }).toMatchSnapshot();
});
