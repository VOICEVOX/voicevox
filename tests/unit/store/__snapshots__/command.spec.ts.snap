// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`コマンド実行で履歴が作られる 1`] = `
{
  "audioKeys": [
    "00000000-0000-4000-0000-000000000001",
  ],
  "redoCommands": {
    "song": [],
    "talk": [],
  },
  "undoCommands": {
    "song": [],
    "talk": [
      {
        "id": "00000000-0000-4000-0000-000000000002",
        "redoPatches": [
          {
            "op": "replace",
            "path": [
              "audioKeys",
            ],
            "value": [
              "00000000-0000-4000-0000-000000000001",
            ],
          },
        ],
        "undoPatches": [
          {
            "op": "replace",
            "path": [
              "audioKeys",
            ],
            "value": [],
          },
        ],
      },
    ],
  },
}
`;
